src/membraneWebRTC.ts

```diff
@@ -86,37 +94,86 @@ export interface SimulcastConfig {
  * Track's context i.e. all data that can be useful when operating on track.
  */
 export interface TrackContext {
-  track: MediaStreamTrack | null;
+  readonly track: MediaStreamTrack | null;
+
   /**
    * Stream this track belongs to.
    */
-  stream: MediaStream | null;
+  readonly stream: MediaStream | null;
+
   /**
    * Peer this track comes from.
    */
-  peer: Peer;
+  readonly peer: Peer;
+
   /**
    * Track id. It is generated by RTC engine and takes form `peer_id:<random_uuidv4>`.
    * It is WebRTC agnostic i.e. it does not contain `mid` or `stream id`.
    */
-  trackId: string;
+
+  readonly trackId: string;
   /**
    * Simulcast configuration.
+   * Only present for local tracks.
    */
-  simulcastConfig: SimulcastConfig;
+  readonly simulcastConfig?: SimulcastConfig;
+
   /**
-   * Any info that was passed in {@link addTrack}.
+   * Any info that was passed in {@link MembraneWebRTC.addTrack}.
    */
-  metadata: any;
+  readonly metadata: any;
 
-  maxBandwidth?: TrackBandwidthLimit;
+  readonly maxBandwidth?: TrackBandwidthLimit;
+
+  /**
+   * Encoding that is currently received.
+   * Only present for remote tracks.
+   */
+  readonly encoding?: TrackEncoding;
+
+  /**
+   * The reason of currently selected encoding.
+   * Only present for remote tracks.
+   */
+  readonly encodingReason?: EncodingReason;
+
+  /**
+   * Called each time track encoding has changed.
+   *
+   * Track encoding can change in the following cases:
+   * - when user requested a change
+   * - when sender stopped sending some encoding (because of bandwidth change)
+   * - when receiver doesn't have enough bandwidth
+   *
+   * Some of those reasons are indicated in {@link TrackContext.encodingReason}.
+   */
+  onEncodingChanged?: (this: TrackContext) => void;
+}
+
+class TrackContextImpl implements TrackContext {
+  peer: Peer;
+  trackId: string;
+  track: MediaStreamTrack | null = null;
+  stream: MediaStream | null = null;
+  metadata: any;
+  simulcastConfig?: SimulcastConfig;
+  maxBandwidth: TrackBandwidthLimit = 0;
+  encoding?: TrackEncoding;
+  encodingReason?: EncodingReason;
+  onEncodingChanged?: (this: TrackContext) => void;
+
+  constructor(peer: Peer, trackId: string, metadata: any) {
+    this.peer = peer;
+    this.trackId = trackId;
+    this.metadata = metadata;
+  }
 }
```

Comment:
Instead of all of that, why not use https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype ? Callbacks would just receive `Readonly<TrackContext>` and we would be done. We can even define another type for that to make it look pretty and this way, we could avoid code duplication.